File: docs/algorithms/code/algorithms/tree/recursion-traversal/preorder-traversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function preorderTraversalHelper(root: TreeNode | null, arr: number[]) {
  if(!root) {
    return;
  }

  arr.push(root.val);
  if(root.left) {
    preorderTraversalHelper(root.left, arr);
  }
  if(root.right) {
    preorderTraversalHelper(root.right, arr);
  }
}

function preorderTraversal(root: TreeNode | null): number[] {
  if(!root) {
    return [];
  }

  const arr: number[] = []
  preorderTraversalHelper(root, arr);
  return arr;
};

File: docs/algorithms/code/algorithms/preorderTraversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func preorderTraversalHelper(root *TreeNode, arr *[]int) {
	if root == nil {
		return;
	}

	*arr = append(*arr, root.Val);

	if root.Left != nil {
		preorderTraversalHelper(root.Left, arr);
	}
	if root.Right != nil {
		preorderTraversalHelper(root.Right, arr);
	}
}
 
func preorderTraversal(root *TreeNode) []int {
	arr := []int{};
 
	if root == nil {
		return arr;
	}
 
	preorderTraversalHelper(root, &arr);
	return arr;
}